### PR TITLE
Feature/aoaf 261 fe create node app including ecs terraform for the presenter account web service

### PR DESCRIPTION
**Pull request by okoye-companieshouse**
- Create a node skeleton app. 
- Dockerise the app. 
- Add ECS terraform.
- Add to new presenter account module in docker-chs-development
- Add health check

### CODE REVIEW

File: terraform/groups/ecs-service/data.tf [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+data "aws_ecs_cluster" "ecs_cluster" {
+  cluster_name = "${local.name_prefix}-cluster"
+}
+
+data "aws_iam_role" "ecs_cluster_iam_role" {
+  name = "${local.name_prefix}-ecs-task-execution-role"
+}
+
+data "aws_lb" "filing_maintain_lb" {
```
Comment (amartin7663) [inline terraform/groups/ecs-service/data.tf:28]: This service will be in the identity cluster so this needs to be
```suggestion
data "aws_lb" "identity_lb" {
```

File: terraform/groups/ecs-service/locals.tf [hunk 2]
```diff
@@ -0,0 +1,84 @@
+# Define all hardcoded local variable and local variables looked up from data resources
+locals {
+  stack_name                = "filing-maintain" # this must match the stack name the service deploys into
```
Comment (amartin7663) [inline terraform/groups/ecs-service/locals.tf:3]: Same here "identity"

File: terraform/groups/ecs-service/main.tf [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+module "secrets" {
+  source = "git@github.com:companieshouse/terraform-modules//aws/ecs/secrets?ref=1.0.192"
+
+  name_prefix = "${local.service_name}-${var.environment}"
+  environment = var.environment
+  kms_key_id  = data.aws_kms_key.kms_key.id
+  secrets     = local.parameter_store_secrets
+}
+
+module "ecs-service" {
```
Comment (amartin7663) [inline terraform/groups/ecs-service/main.tf:31]: You're missing some stuff in here, such as cloudwatch warning etc.. Have a look here: https://github.com/companieshouse/presenter-account-api/blob/main/terraform/groups/ecs-service/main.tf
You don't need the ERIC stuff though, as that's API only. Also you should use the latest version of the ecs-service release

**Review by amartin7663 — CHANGES_REQUESTED** [review]
You also need a version file like this: https://github.com/companieshouse/presenter-account-api/pull/3

File: Tiltfile.dev [hunk 4]
```diff
@@ -0,0 +1,32 @@
+custom_build(
+    ref = '416670754337.dkr.ecr.eu-west-2.amazonaws.com/presenter-account-web',
+    #the following build-command was updated as specified by https://github.com/companieshouse/docker-chs-development/pull/581
+    command = 'DOCKER_BUILDKIT=0 docker build --build-arg SSH_PRIVATE_KEY="$(cat ~/.ssh/id_rsa)" --build-arg SSH_PRIVATE_KEY_PASSPHRASE --tag $EXPECTED_REF .',
+    live_update = [
```
Comment (HarryBosworth) [inline Tiltfile.dev:5]:
```suggestion
    command = 'DOCKER_BUILDKIT=0 docker build --build-arg SSH_PRIVATE_KEY="$(ssh_key_path="$(ssh -G github.com | grep -e '^identityfile.' | head -n1 | sed 's|^identityfile \(.\)|\1|')"; if [ -z "${ssh_key_path}" ]; then echo "Could not find ssh key path for github.com">&2; false; elif [ -f "${ssh_key_path}" ]; then cat "${ssh_key_path}"; else echo "Could not find ssh key for github at ${ssh_key_path}" >&2; false; fi)" --build-arg SSH_PRIVATE_KEY_PASSPHRASE --tag $EXPECTED_REF .',
    live_update = [
```

We should be using the command from the readme for the ssh private key.

File: package.json
Comment (HarryBosworth) [inline package.json:1]: `docker_start.sh` uses the npm script `dev`. This script will need to be added into here.
Something like this:

`"dev": "nodemon --signal SIGTERM --watch src --ignore 'src/**/*.unit.ts' --ext ts --exec 'node --require ts-node/register --inspect=0.0.0.0:9229 ./src/server.ts'",`

If you use this then you also need to check `ts-node` is installed as a dev dependency.

File: sonar-project.properties [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+
+sonar.host.url=${env.CODE_ANALYSIS_HOST_URL}
+sonar.login=${env.CODE_ANALYSIS_LOGIN}
+sonar.password=${env.CODE_ANALYSIS_PASSWORD}
+sonar.sources=src
+
+sonar.exclusions=**/bin/**, **/dist/**, **/coverage/**, **/node_modules/**
+sonar.tests=test
+sonar.test.inclusions=**/test/**.spec.*
+sonar.typescript.lcov.reportPaths=coverage/lcov.info
```
Comment (HarryBosworth) [inline sonar-project.properties:12]:
```suggestion
sonar.typescript.lcov.reportPaths=coverage/lcov.info

```

I think this file needs an empty line on the end due to weirdness arround hosw the pipeline injects properties.

File: src/routers/handlers/generic.ts [hunk 6]
```diff
@@ -1,17 +1,38 @@
...
+import errorManifest from "../../lib/utils/error_manifests/default";
+import { Request } from "express";
 
-export class GenericHandler {
+export interface BaseViewData {
+    errors: any
+    title: string
+    isSignedIn: boolean
+    backURL: string | null
+    sampleKey?: string
```
Comment (HarryBosworth) [inline src/routers/handlers/generic.ts:12]: What is this variable? Is it needed?

File: src/utils/constants/home.ts
Comment (HarryBosworth) [inline src/utils/constants/home.ts:1]: It doesn't make sense to define these here.
They should be defined in the HomeHandler.

File: src/router.dispatch.ts [hunk 7]
```diff
@@ -1,16 +1,19 @@
...
-    app.use("*", (req: Request, res: Response) => {
-        res.status(404).render("partials/error_400");
-    });
+
+    const router = Router();
+    app.use(servicePathPrefix, router);
+
+    router.use("/", HomeRouter);
+    router.use(healthcheckUrl, HealthCheckRouter);
+
```
Comment (HarryBosworth) [inline src/router.dispatch.ts:14]: We should add the session and auth middleware now. It will save us many headaches down the road.
Make sure they are attched to the router after the Home, and healthcheck endpoints so it won't be active on those pages. 
```suggestion
    router.use('/', sessionMiddleware)
    router.use('/', authMiddleware)
```

**Review by HarryBosworth — CHANGES_REQUESTED** [review]
You should also link the docker-chs-development PR tothis PR since they are so tightly coupled.

File: src/routers/handlers/generic.ts (hunk 6 above)
Comment (okoye-companieshouse) [reply src/routers/handlers/generic.ts:12]: thought of using a string[] to hold placeholder values but can take it out

File: terraform/groups/ecs-service/profiles/staging-eu-west-2/staging/vars [hunk 8]
```diff
@@ -0,0 +1,13 @@
+environment = "staging"
```
Comment (amartin7663) [inline terraform/groups/ecs-service/profiles/staging-eu-west-2/staging/vars:1]: In here you should add the service scale up cron job tasks like this: https://github.com/companieshouse/presenter-account-api/blob/main/terraform/groups/ecs-service/profiles/staging-eu-west-2/staging/vars

File: terraform/groups/ecs-service/locals.tf [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+    for sec in data.aws_ssm_parameter.secret :
+    trimprefix(sec.name, "/${local.name_prefix}/") => sec.arn
+  }
+
+  service_secrets_arn_map = {
+    for sec in module.secrets.secrets :
+    trimprefix(sec.name, "/${local.service_name}-${var.environment}/") => sec.arn
+  }
+
+  task_secrets = [
```
Comment (amartin7663) [inline terraform/groups/ecs-service/locals.tf:58]: You don't need to list all the secrets and environment variables here anymore you can add them programmatically like this: https://github.com/companieshouse/presenter-account-consumer/blob/main/terraform/groups/ecs-service/locals.tf

File: terraform/groups/ecs-service/data.tf [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+data "aws_ecs_cluster" "ecs_cluster" {
+  cluster_name = "${local.name_prefix}-cluster"
+}
+
+data "aws_iam_role" "ecs_cluster_iam_role" {
+  name = "${local.name_prefix}-ecs-task-execution-role"
+}
+
+data "aws_lb" "identity_lb" {
+  name = "${var.environment}-chs-internalapi"
```
Comment (amartin7663) [inline terraform/groups/ecs-service/data.tf:37]: This should be the chs chgovuk load balancer as it's a web service, not an api

**Review by amartin7663 — COMMENTED** [review]
Terraform looks good now 👍

File: src/Cache.ts
Comment (HarryBosworth) [inline src/Cache.ts:1]: Is this file needed?

File: src/app.ts [hunk 11]
```diff
@@ -1,25 +1,28 @@
...
 const app = express();
 
 // const viewPath = path.join(__dirname, "/views");
 app.set("views", [
     path.join(__dirname, "/views"),
-    path.join(__dirname, "/../node_modules/govuk-frontend")
+    path.join(__dirname, "/../node_modules/govuk-frontend"),
+    path.join(__dirname, "../node_modules/govuk-frontend"), // This if for when using ts-node since the working directory is src
+    path.join(__dirname, "node_modules/govuk-frontend/components")
 ]);
```
Comment (HarryBosworth) [inline src/app.ts:16]: These template paths don't work and result in an error when accessing the service.
```suggestion
app.set("views", [
    path.join(__dirname, "views"),
    path.join(__dirname, "node_modules/govuk-frontend"),
    path.join(__dirname, "../node_modules/govuk-frontend"), // This if for when using ts-node since the working directory is src
    path.join(__dirname, "node_modules/govuk-frontend/components")
]);
```

**Review by HarryBosworth — CHANGES_REQUESTED** [review]
Also, the static assets are failing to build with the error

```
> gulp build --sourcemap

[10:12:56] Using gulpfile ~/dev/docker-chs-development/repositories/presenter-account-web/gulpfile.js
[10:12:56] Starting 'build'...
[10:12:56] Starting 'clean'...
[10:12:56] 'clean' errored after 272 ms
[10:12:56] Error: ENOENT: no such file or directory, scandir '/Users/hbosworth/dev/docker-chs-development/repositories/presenter-account-web/assets/public/css'
[10:12:56] 'build' errored after 274 ms
Failed to build assets
```

File: src/Cache.ts
Comment (okoye-companieshouse) [reply src/Cache.ts:1]: No no took it out

File: Tiltfile.dev (hunk 4 above)
Comment (HarryBosworth) [reply Tiltfile.dev:5]: This still needs to be resolved

File: routes.yaml
Comment (HarryBosworth) [inline routes.yaml:1]: This file isn't needed since we are deploying to ECS.
It should be removed

File: sonar-project.properties (hunk 5 above)
Comment (HarryBosworth) [reply sonar-project.properties:12]: This also looks like it is not reolved. 
When I view the view in the PR the trailing newline is not there.
However, when I check it out locally it is. That's strange. 
You should try adding this trailing new line and if it still doesn't work, just leave it.

File: test/routers/home.unit.ts
Comment (HarryBosworth) [inline test/routers/home.unit.ts:1]: This test is failing.

**Review by HarryBosworth — CHANGES_REQUESTED** [review]

**Review by HarryBosworth — CHANGES_REQUESTED** [review]
Everything looks good, except the test which is still failing.

```
 FAIL  test/routers/home.unit.ts (8.356 s)
  home page tests
    ✕ should render the home page (65 ms)

  ● home page tests › should render the home page
  ```

**Review by HarryBosworth — APPROVED** [review]

**Review by amartin7663 — APPROVED** [review]
